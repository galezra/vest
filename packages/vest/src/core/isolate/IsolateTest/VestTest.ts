import { Maybe, invariant, isPromise, optionalFunctionValue } from 'vest-utils';
import { IsolateMutator, IsolateSelectors, TIsolate } from 'vestjs-runtime';

import { ErrorStrings } from 'ErrorStrings';
import type { TIsolateTest } from 'IsolateTest';
import {
  IsolateTestStateMachine,
  TestAction,
  TestStatus,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestIsolate } from 'VestIsolate';
import { VestIsolateType } from 'VestIsolateType';

export class VestTest extends VestIsolate {
  static stateMachine = IsolateTestStateMachine;

  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName,
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static is(isolate?: Maybe<TIsolate>): isolate is TIsolateTest {
    return IsolateSelectors.isIsolateType<TIsolateTest>(
      isolate,
      VestIsolateType.Test,
    );
  }

  static isX(isolate?: Maybe<TIsolate>): asserts isolate is TIsolateTest {
    invariant(VestTest.is(isolate), ErrorStrings.EXPECTED_VEST_TEST);
  }

  static cast<F extends TFieldName = string, G extends TGroupName = string>(
    isolate?: Maybe<TIsolate>,
  ): TIsolateTest<F, G> {
    VestTest.isX(isolate);
    return isolate as TIsolateTest<F, G>;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  // Mutate

  // static setPending(test: TIsolateTest) {
  //   this.setStatus(test, TestStatus.PENDING);
  // }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED,
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data'],
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    IsolateMutator.abort(test, TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }
}
