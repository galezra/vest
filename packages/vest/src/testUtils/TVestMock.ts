import { TFieldName, TGroupName } from 'SuiteResultTypes';
import * as vest from 'vest';

export type TVestMock = typeof vest;

export type TTestSuiteCallback = (..._args: any[]) => void;
export type TTestSuite = vest.Suite<TFieldName, TGroupName, TTestSuiteCallback>;
