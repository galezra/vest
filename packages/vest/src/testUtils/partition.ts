export default function partition<T>(
  array: T[],
  predicate: (value: T, _index: number, array: T[]) => boolean,
): [T[], T[]] {
  return array.reduce(
    (partitions: [T[], T[]], value, number) => {
      partitions[predicate(value, number, array) ? 0 : 1].push(value);
      return partitions;
    },
    [[], []],
  );
}
