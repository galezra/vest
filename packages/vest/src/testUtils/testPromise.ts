export function TestPromise(cb: (_done: () => void) => void): Promise<void> {
  return new Promise<void>(done => cb(done));
}
