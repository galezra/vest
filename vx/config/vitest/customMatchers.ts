import { isFunction } from 'vest-utils';

expect.extend({
  isDeepCopyOf,
});

// eslint-disable-next-line complexity
function isDeepCopyOf(
  source: any,
  clone: any,
): { pass: boolean; message: () => string } {
  const queue = [[source, clone]];

  outer: while (queue.length) {
    const [source, clone] = (queue as any[]).shift();

    if (!source || typeof source !== 'object') {
      if (!isFunction(clone)) {
        if (clone !== source) {
          return {
            pass: false,
            message: () => `Source and clone are not identical`,
          };
        }
      }
      continue;
    }

    if (clone === source) {
      return {
        pass: false,
        message: () =>
          `Source and clone are the same object. Expected a deep copy. ${JSON.stringify(
            source,
          )}===${JSON.stringify(clone)}`,
      };
    }

    // Short circuit
    if (
      (clone && !source) ||
      (source && !clone) ||
      typeof source !== typeof clone
    ) {
      return {
        pass: false,
        message: () =>
          `Source and clone are not of the same type: ${JSON.stringify(
            source,
          )} does not equal ${JSON.stringify(clone)}`,
      };
    }

    if (Array.isArray(source)) {
      // Short circuit
      if (!Array.isArray(clone) || source.length !== clone.length) {
        return {
          pass: false,
          message: () =>
            `source and clone arrays are not identical. ${JSON.stringify(
              source,
            )} does not equal ${JSON.stringify(clone)}`,
        };
      }
      source.forEach((_, i) => {
        queue.push([source[i], clone[i]]);
      });
    } else if (typeof source === 'object') {
      Object.keys(source).forEach(key => queue.push([source[key], clone[key]]));
    }

    continue outer;
  }

  return { pass: true, message: () => 'success' };
}
