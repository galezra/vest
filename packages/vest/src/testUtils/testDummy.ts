import { faker } from '@faker-js/faker';
import { vi } from 'vitest';

import { test as vestTest, warn } from 'vest';

/**
 * Generates dummy vest tests.
 */
// eslint-disable-next-line max-lines-per-function
const testDummy = () => {
  const failing = (
    name: string = faker.lorem.word(),
    message: string = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        throw new Error();
      }),
    );

    return to;
  };

  const failingWarning = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        throw new Error();
      }),
    );

    return to;
  };

  const passing = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(name, message, vi.fn());

    return to;
  };

  const passingWarning = (
    name = faker.lorem.word(),
    message = faker.lorem.words(),
  ) => {
    const to = vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
      }),
    );
    return to;
  };

  const failingAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(
        () =>
          new Promise((_, reject) => {
            setTimeout(reject, time);
          }),
      ),
    );

  const failingWarningAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        return new Promise((_, reject) => {
          setTimeout(reject, time);
        });
      }),
    );

  const passingAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(
        () =>
          new Promise(resolve => {
            setTimeout(resolve, time);
          }),
      ),
    );

  const passingWarningAsync = (
    name = faker.lorem.word(),
    { message = faker.lorem.words(), time = 0 } = {},
  ) =>
    vestTest(
      name,
      message,
      vi.fn(() => {
        warn();
        return new Promise(resolve => {
          setTimeout(resolve, time);
        });
      }),
    );

  return {
    failing,
    failingAsync,
    failingWarning,
    failingWarningAsync,
    passing,
    passingAsync,
    passingWarning,
    passingWarningAsync,
  };
};

export const dummyTest = testDummy();

export type TDummyTest = typeof dummyTest;
