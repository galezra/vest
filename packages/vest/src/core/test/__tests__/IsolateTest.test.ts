import { TIsolateTest } from 'IsolateTest';
import { VestTest } from 'VestTest';
import { mockIsolateTest } from 'vestMocks';
import { describe, it, expect, beforeEach, test, vi } from 'vitest';
import wait from 'wait';

import { TestPromise } from '../../../testUtils/testPromise';

import * as vest from 'vest';

const fieldName = 'unicycle';
const message = 'I am Root.';

describe('IsolateTest', () => {
  let testObject: TIsolateTest;

  beforeEach(() => {
    testObject = mockIsolateTest({
      fieldName,
      message,
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  describe('testObject.warn', () => {
    it('Should mark the test as warning', () => {
      expect(VestTest.warns(testObject)).toBe(false);
      VestTest.warn(testObject);
      expect(VestTest.warns(testObject)).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      testObject = mockIsolateTest({ fieldName, message });
    });

    it('Should set status to failed', () => {
      expect(VestTest.isFailing(testObject)).toBe(false);
      VestTest.fail(testObject);
      expect(VestTest.isFailing(testObject)).toBe(true);
    });
  });

  describe('testObject.cancel', () => {
    it('Should set the testObject to cancel', () => {
      let testObject: TIsolateTest;
      return TestPromise(done => {
        const suite = vest.create(() => {
          testObject = vest.test('f1', async () => {
            await wait(100);
          });
          vest.test('f2', async () => {
            await wait(100);
          });
          VestTest.cancel(testObject);
        });
        suite();

        expect(VestTest.isCanceled(testObject)).toBe(true);
        done();
      });
    });

    describe('final statuses', () => {
      const control = vi.fn();
      beforeEach(() => {
        vi.resetAllMocks();
      });
      it('keep status unchanged when `failed`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          VestTest.fail(testObject);
          expect(VestTest.isFailing(testObject)).toBe(true);
          VestTest.skip(testObject);
          expect(VestTest.isSkipped(testObject)).toBe(false);
          expect(VestTest.isFailing(testObject)).toBe(true);
          VestTest.cancel(testObject);
          expect(VestTest.isCanceled(testObject)).toBe(false);
          expect(VestTest.isFailing(testObject)).toBe(true);
          VestTest.setPending(testObject);
          expect(VestTest.isPending(testObject)).toBe(false);
          expect(VestTest.isFailing(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });

      it('keep status unchanged when `canceled`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          VestTest.cancel(testObject);
          expect(VestTest.isCanceled(testObject)).toBe(true);
          VestTest.fail(testObject);
          expect(VestTest.isCanceled(testObject)).toBe(true);
          expect(VestTest.isFailing(testObject)).toBe(false);
          VestTest.skip(testObject);
          expect(VestTest.isSkipped(testObject)).toBe(false);
          expect(VestTest.isCanceled(testObject)).toBe(true);
          VestTest.setPending(testObject);
          expect(VestTest.isPending(testObject)).toBe(false);
          expect(VestTest.isCanceled(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });
    });
  });
});
