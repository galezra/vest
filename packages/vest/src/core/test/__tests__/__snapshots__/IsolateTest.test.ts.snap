// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IsolateTest > TestObject constructor 1`] = `
{
  "$type": "UnitTest",
  "children": [],
  "data": {
    "fieldName": "unicycle",
    "message": "I am Root.",
    "severity": "error",
    "testFn": [MockFunction spy],
    "type": "Test",
  },
  "key": null,
  "keys": {},
  "output": null,
  "parent": null,
  "status": "UNTESTED",
}
`;

exports[`IsolateTest > testObject.warn > Should mark the test as warning 1`] = `
{
  "$type": "UnitTest",
  "children": [],
  "data": {
    "fieldName": "unicycle",
    "message": "I am Root.",
    "severity": "warning",
    "testFn": [MockFunction spy],
    "type": "Test",
  },
  "key": null,
  "keys": {},
  "output": null,
  "parent": null,
  "status": "UNTESTED",
}
`;
